src/NuGet.Clients/NuGet.CommandLine/Commands/DeprecatedCommandAttribute.cs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Reflection;
+
+namespace NuGet.CommandLine
+{
+    /// <summary>
+    /// Attribute for marking nuget.exe commands as deprecated.
+    ///
+    /// By using this attribute in ICommand-based classes, it will show a warning message each
+    /// command invocation and in each help command invocation.
+    /// 
+    /// <see cref="ICommand"/>
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = false, Inherited = false)]
+    public sealed class DeprecatedCommandAttribute : Attribute
+    {
+        /// <summary>
+        /// The concrete type that is an alternative to the command that has this DeprecatedCommand attribute.
+        /// If no alternative is provided in the attribute, this property will be null
+        /// </summary>
+        public Type AlternativeCommand { get; set; }
+
+        public DeprecatedCommandAttribute()
+        {
+            AlternativeCommand = null;
+        }
+
+        public DeprecatedCommandAttribute(Type alternativeCommand)
+        {
+            if (!typeof(ICommand).IsAssignableFrom(alternativeCommand))
+            {
+                var msg = string.Format("{0} must be a {1} that implements {2} interface",
+                    nameof(alternativeCommand),
+                    nameof(Type),
+                    typeof(ICommand).FullName);
+                throw new ArgumentException(msg);
+            }
+            AlternativeCommand = alternativeCommand;
+        }
+
+        /// <summary>
+        /// Generates the deprecation warning message
+        /// </summary>
+        /// <param name="currentCommand">name of the current command that is deprecated</param>
+        /// <returns>The deprecation warning message</returns>
+        public string GetDeprecationMessage(string currentCommand)
+        {
+            var binaryName = Assembly.GetExecutingAssembly().GetName().Name;
```

Comment:
What's the point of the binary name? 

You're getting nuget here? 
I don't think that necessarily needs to be in the message. 

Do we have a pattern doing this or?

Reply:
> What's the point of the binary name?

Get the .exe name for the message.

> You're getting nuget here?

Yes. I'm getting the name of the .exe; see last question

> I don't think that necessarily needs to be in the message.

If that's the case, the message would look like this:

```
WARNING: 'greet' is deprecated. Use 'hello' instead
```

The message above, seems out of context.

Also, I don't want to hardcode the nuget name.

> Do we have a pattern doing this or?
I was inspired by the help command:
https://github.com/NuGet/NuGet.Client/blob/132a4493d44ba9c4ae2e7cfe5a9efd1538c7fc1a/src/NuGet.Clients/NuGet.CommandLine/Commands/HelpCommand.cs#L42

Reply:
You could say command, but it's fine as is.

Reply:
I have `nuget.4.9.4.exe`, `nuget.5.0.0.exe` and others on my path, but no `nuget.exe`. This approach correctly tells me the command, unlike hardcoding "nuget", which is what I think Nikolche's first comment asked about. I like it this way, it feels more polished.

Reply:
My bad. This only always returns 'NuGet'. It doesn't return the .exe name.

Reply:
I don't think we need the exe part. 
Especialyl if other commands do it without. 

For reference, i think `CodeBase` will provide the name of the assembly extension and all (path including :) )

Reply:
Ok. I will keep it in that way.